File: XiePinyin/Logic/Composer.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

namespace XiePinyin.Logic
{
    public class Composer
    {
        readonly Pinyin pinyin;
        readonly List<CharReading> readingsSimp = new List<CharReading>();
        readonly List<CharReading> readingsTrad = new List<CharReading>();

        public Composer(string sourcesFolder)
        {
            pinyin = new Pinyin(Path.Combine(sourcesFolder, "pinyin.txt"));
            JsonSerializer ser = new JsonSerializer();
            using (StreamReader sr = new StreamReader(Path.Combine("wwwroot", "simp-map.json")))
            {
                readingsSimp = ser.Deserialize(sr, typeof(List<CharReading>)) as List<CharReading>;
            }
            using (StreamReader sr = new StreamReader(Path.Combine("wwwroot", "trad-map.json")))
            {
                readingsTrad = ser.Deserialize(sr, typeof(List<CharReading>)) as List<CharReading>;
            }
            //addPunctReadings(readingsSimp, true);
            //addPunctReadings(readingsTrad, false);
        }

        void addPunctReadings(List<CharReading> readings, bool simp)
        {
            readings.Add(new CharReading { Hanzi = "。", Pinyin = "." });
            readings.Add(new CharReading { Hanzi = "·", Pinyin = "." });
            readings.Add(new CharReading { Hanzi = "，", Pinyin = "," });
            readings.Add(new CharReading { Hanzi = "、", Pinyin = "," });
            readings.Add(new CharReading { Hanzi = "？", Pinyin = "?" });
            readings.Add(new CharReading { Hanzi = "！", Pinyin = "!" });
            readings.Add(new CharReading { Hanzi = "：", Pinyin = ":" });
            readings.Add(new CharReading { Hanzi = "；", Pinyin = ";" });
            readings.Add(new CharReading { Hanzi = "……", Pinyin = ". . ." });
            readings.Add(new CharReading { Hanzi = "…", Pinyin = ". . ." });
            readings.Add(new CharReading { Hanzi = "【", Pinyin = "(" });
            readings.Add(new CharReading { Hanzi = "（", Pinyin = "(" });
            readings.Add(new CharReading { Hanzi = "】", Pinyin = ")" });
            readings.Add(new CharReading { Hanzi = "）", Pinyin = ")" });
            readings.Add(new CharReading { Hanzi = "《", Pinyin = "(" });
            readings.Add(new CharReading { Hanzi = "》", Pinyin = ")" });
            readings.Add(new CharReading { Hanzi = "——", Pinyin = "- -" });
            if (simp)
            {
                readings.Add(new CharReading { Hanzi = "“", Pinyin = "\"" });
                readings.Add(new CharReading { Hanzi = "”", Pinyin = "\"" });
                readings.Add(new CharReading { Hanzi = "‘", Pinyin = "'" });
                readings.Add(new CharReading { Hanzi = "’", Pinyin = "'" });
            }
            else
            {
                readings.Add(new CharReading { Hanzi = " 「", Pinyin = "\"" });
                readings.Add(new CharReading { Hanzi = "」", Pinyin = "\"" });
                readings.Add(new CharReading { Hanzi = "『", Pinyin = "'" });
                readings.Add(new CharReading { Hanzi = "』", Pinyin = "'" });
            }
        }

        public List<List<string>> Resolve(string pinyinInput, bool isSimp, out List<string> pinyinSylls)
        {
            var res = new List<List<string>>();
            List<CharReading> readings = isSimp ? readingsSimp : readingsTrad;
            string pinyinInputLo = pinyinInput.ToLowerInvariant();
            var loSylls = pinyin.SplitSyllables(pinyinInputLo);
            string loSyllsConcat = "";
            for (int i = 0; i < loSylls.Count; ++i) { if (i != 0) loSyllsConcat += ' '; loSyllsConcat += loSylls[i]; }
            foreach (var r in readings)
            {
                if (r.Pinyin == loSyllsConcat)
                {
                    var itm = new List<string>();
                    itm.Add(r.Hanzi);
                    res.Add(itm);
                }
            }
            pinyinSylls = getOrigSylls(pinyinInput, pinyinInputLo, loSylls);
            return res;
        }

        List<string> getOrigSylls(string orig, string lo, List<string> loSylls)
        {
            var res = new List<string>();
            int ix = 0;
            for (int i = 0; i < loSylls.Count; ++i)
            {
                ix = lo.IndexOf(loSylls[i], ix);
                string origSyll = orig.Substring(ix, loSylls[i].Length);
                ix += origSyll.Length;
                res.Add(origSyll);
            }
            return res;
        }

        public string PinyinNumsToSurf(string pyNums)
        {
            string pyNumsLo = pyNums.ToLowerInvariant();
            var loSylls = pinyin.SplitSyllables(pyNumsLo);
            var loSyllsPretty = new List<string>();
            foreach (var ls in loSylls)
            {
                string pretty = pinyin.NumsToSurf(ls);
                if (pretty == null) pretty = ls;
                loSyllsPretty.Add(pretty);
            }
            var origSylls = getOrigSylls(pyNums, pyNumsLo, loSylls);
            string res = "";
            for (int i = 0; i < loSyllsPretty.Count; ++i)
            {
                string loSyllPretty = loSyllsPretty[i];
                if (loSylls[i] == origSylls[i]) res += loSyllPretty;
                else
                {
                    string casedPretty = char.ToUpperInvariant(loSyllPretty[0]).ToString();
                    if (loSyllPretty.Length > 1) casedPretty += loSyllPretty.Substring(1);
                    res += casedPretty;
                }
            }
            return res;
        }
    }
}

File: xiep/internal/logic/composer.go
package logic

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"strings"
	"unicode"
)

type charReading struct {
	Hanzi  string `json:"hanzi"`
	Pinyin string `json:"pinyin"`
}

type composer struct {
	pinyin       *pinyin
	readingsSimp []charReading
	readingsTrad []charReading
}

func loadComposerFromFiles(dataDir string) *composer {
	var res composer
	res.pinyin = loadPinyin()
	res.readingsSimp = loadCharReadings(path.Join(dataDir, "simp-map.json"))
	res.readingsTrad = loadCharReadings(path.Join(dataDir, "trad-map.json"))
	return &res
}

func loadComposerFromString(simpJson, tradJson string) *composer {
	var res composer
	res.pinyin = loadPinyin()
	if e := json.Unmarshal([]byte(simpJson), &res.readingsSimp); e != nil {
		panic(fmt.Sprintf("Error parsing Json: %v", e))
	}
	if e := json.Unmarshal([]byte(tradJson), &res.readingsTrad); e != nil {
		panic(fmt.Sprintf("Error parsing Json: %v", e))
	}
	return &res
}

func loadCharReadings(fnJson string) []charReading {
	f, err := os.Open(fnJson)
	if err != nil {
		panic(fmt.Sprintf("Failed to open file: %v", fnJson))
	}
	//goland:noinspection GoUnhandledErrorResult
	defer f.Close()
	data, err := ioutil.ReadAll(f)
	if err != nil {
		panic(fmt.Sprintf("Failed to read file: %v", fnJson))
	}
	res := make([]charReading, 0)
	if e := json.Unmarshal(data, &res); e != nil {
		panic(fmt.Sprintf("Error parsing Json: %v", e))
	}
	return res
}

// Splits input into pinyin syllables, and returns Hanzi words matching the whole input.
func (cp *composer) Resolve(pinyinInput string, isSimp bool) (pinyinSylls []string, readings [][]string) {
	readings = make([][]string, 0)
	charReadings := cp.readingsTrad
	if isSimp {
		charReadings = cp.readingsSimp
	}
	pinyinInputLo := strings.ToLower(pinyinInput)
	loSylls := cp.pinyin.splitSyllables(pinyinInputLo)
	loSyllsConcat := ""
	for i, syll := range loSylls {
		if i != 0 {
			loSyllsConcat += " "
		}
		loSyllsConcat += syll
	}
	for _, r := range charReadings {
		if r.Pinyin != loSyllsConcat {
			continue
		}
		itm := make([]string, 0, 1)
		itm = append(itm, r.Hanzi)
		readings = append(readings, itm)
	}
	pinyinSylls = getOrigSylls(pinyinInput, pinyinInputLo, loSylls)
	return
}

func getOrigSylls(orig string, lo string, loSylls []string) (origSylls []string) {
	origSylls = make([]string, 0, len(loSylls))
	ix := 0
	for _, loSyll := range loSylls {
		ix += strings.Index(lo[ix:], loSyll)
		origSyll := orig[ix : ix+len(loSyll)]
		origSylls = append(origSylls, origSyll)
	}
	return
}

// Parses input string with numbers, and returns pinyin with pretty tone marks.
func (cp *composer) PinyinNumsToSurf(pyNums string) string {
	pyNumsLo := strings.ToLower(pyNums)
	loSylls := cp.pinyin.splitSyllables(pyNumsLo)
	var loSyllsPretty = make([]string, 0)
	for _, ls := range loSylls {
		pretty := cp.pinyin.numToSurf(ls)
		if pretty == "" {
			pretty = ls
		}
		loSyllsPretty = append(loSyllsPretty, pretty)
	}
	origSylls := getOrigSylls(pyNums, pyNumsLo, loSylls)
	var sb strings.Builder
	for i, loSyllPretty := range loSyllsPretty {
		// Original was lower case
		if loSylls[i] == origSylls[i] {
			sb.WriteString(loSyllPretty)
			continue
		}
		// Make first letter upper case, copy rest as is
		for j, chr := range loSyllPretty {
			if j == 0 {
				sb.WriteRune(unicode.ToUpper(chr))
			} else {
				sb.WriteRune(chr)
			}
		}
	}
	return sb.String()
}
